File: packages/orchestra/src/modules/shells/bar_count.ts
import type { DatasetProfile } from "@/modules/csv/profiler";
import type {
    AnalysisType,
    EncodingMapType,
    FieldType,
    ShellParamsType,
    ShellType,
    ValidationResult,
} from "@/modules/shells/types";

const VEGA_LITE_SCHEMA = "https://vega.github.io/schema/vega-lite/v5.json";

export class BarCountShell implements ShellType {
    analysis: AnalysisType | undefined = undefined;
    name = "Bar Count";
    optional: EncodingMapType = {
        color: { type: "nominal" },
        tooltip: { type: "any" },
    };
    required: EncodingMapType = {
        x: { type: "any" },
    };
    rowSemantics = "aggregate" as const;
    signatures: FieldType[][] = [["nominal"], ["ordinal"], ["quantitative"]];

    compile(params: ShellParamsType, values: Record<string, unknown>[], renderer: "vega-lite"): unknown {
        if (renderer !== "vega-lite") {
            throw new Error("Unsupported renderer");
        }

        const encoding: Record<string, any> = {
            x: {
                field: params.x,
                type: "nominal",
            },
            y: {
                aggregate: "count",
                type: "quantitative",
            },
        };

        if (params.color) {
            encoding.color = {
                field: params.color,
                type: "nominal",
            };
        }

        if (params.tooltip) {
            encoding.tooltip = {
                field: params.tooltip,
            };
        }

        return {
            $schema: VEGA_LITE_SCHEMA,
            data: { values },
            encoding,
            mark: { type: "bar" },
        };
    }

    validate(params: ShellParamsType, profile: DatasetProfile): ValidationResult {
        const xField = params.x;
        if (!xField) {
            return {
                errors: [{ code: "missing_required_encoding" }],
                ok: false,
                warnings: [],
            };
        }

        if (!profile.fields[xField]) {
            return {
                errors: [{ code: "unknown_field" }],
                ok: false,
                warnings: [],
            };
        }

        return {
            errors: [],
            ok: true,
            warnings: [],
        };
    }
}

File: packages/vintent/vintent/vintent/core/modules/templates/bar_count.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Literal, Optional, Union

from .types import (
    VEGA_LITE_SCHEMA,
    AnalysisType,
    DatasetProfile,
    EncodingMapType,
    FieldType,
    RendererType,
    ShellParamsType,
    ValidationResult,
)


@dataclass
class BarCountShell:
    """Bar Count shell implementation for counting occurrences."""

    # ShellType interface implementation
    analysis: Optional[AnalysisType] = None
    name: str = "Bar Count"
    description: Optional[str] = None
    signatures: List[List[FieldType]] = field(default_factory=lambda: [["nominal"], ["ordinal"], ["quantitative"]])
    required: EncodingMapType = field(default_factory=lambda: {"x": {"type": "any"}})
    optional: Union[EncodingMapType, Literal["any"], None] = field(
        default_factory=lambda: {"color": {"type": "nominal"}, "tooltip": {"type": "any"}}
    )
    rowSemantics: Literal["rowwise", "aggregate"] = "aggregate"

    def compile(
        self, params: ShellParamsType, values: List[Dict[str, Any]], renderer: RendererType = "vega-lite"
    ) -> Dict[str, Any]:
        """Compile shell parameters into a Vega-Lite specification."""
        if renderer != "vega-lite":
            raise ValueError("Unsupported renderer")

        encoding: Dict[str, Any] = {
            "x": {"field": params.get("x", ""), "type": "nominal"},
            "y": {"aggregate": "count", "type": "quantitative"},
        }

        # Add optional color encoding if provided
        if params.get("color"):
            encoding["color"] = {"field": params["color"], "type": "nominal"}

        # Add optional tooltip encoding if provided
        if params.get("tooltip"):
            encoding["tooltip"] = {"field": params["tooltip"]}

        return {"$schema": VEGA_LITE_SCHEMA, "data": {"values": values}, "encoding": encoding, "mark": {"type": "bar"}}

    def validate(self, params: ShellParamsType, profile: DatasetProfile) -> ValidationResult:
        """Validate parameters against dataset profile."""
        x_field = params.get("x")

        # Check required field is present
        if not x_field:
            return {"errors": [{"code": "missing_required_encoding"}], "ok": False, "warnings": []}

        # Check if field exists in profile
        if x_field not in profile.get("fields", {}):
            return {"errors": [{"code": "unknown_field"}], "ok": False, "warnings": []}

        return {"errors": [], "ok": True, "warnings": []}
